solo-io/gloo
# Fix Istio upstream upgrade test

# Description

Flake was reported on `IstioIntegration/TestUpgrade` test. I was able to reproduce it locally- the wrong Upstream getting applied as part of the test: https://github.com/solo-io/gloo/blob/main/test/kubernetes/e2e/features/istio/gloo_gateway_auto_mtls_suite.go#L196 It should be the sslConfig upstream, but it's applying the disabled auto mtls upstream. Since it is passing in ci, it looks like there is test pollution between the IstioIntegration automtls tests. 

This could be caused by the upstream resources having the same name. So when the resources are applied there is an error on the VirtualService until the upstream is created: 
```
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"gateway.solo.io/v1","kind":"VirtualService","metadata":{"annotations":{},"name":"httpbin-vs","namespace":"automtls-istio-edge-api-test"},"spec":{"virtualHost":{"domains":["httpbin"],"routes":[{"matchers":[{"prefix":"/"}],"routeAction":{"single":{"upstream":{"name":"httpbin-upstream","namespace":"automtls-istio-edge-api-test"}}}}]}}}
  creationTimestamp: "2024-05-29T19:21:48Z"
  generation: 4
  name: httpbin-vs
  namespace: automtls-istio-edge-api-test
  resourceVersion: "15212"
  uid: 15c94fb5-3691-49d8-ada5-e1b1327b2289
spec:
  virtualHost:
    domains:
    - httpbin
    routes:
    - matchers:
      - prefix: /
      routeAction:
        single:
          upstream:
            name: httpbin-upstream
            namespace: automtls-istio-edge-api-test
status:
  statuses:
    automtls-istio-edge-api-test:
      reportedBy: gloo
      state: Accepted
      subresourceStatuses:
        '*v1.Proxy.gateway-proxy_automtls-istio-edge-api-test':
          reason: "warning: \n  Route Warning: InvalidDestinationWarning. Reason:
            *v1.Upstream { automtls-istio-edge-api-test.httpbin-upstream } not found"
          reportedBy: gloo
          state: Warning
```

When the upstream is present, the status correctly updates:
```
status:
  statuses:
    automtls-istio-edge-api-test:
      reportedBy: gloo
      state: Accepted
      subresourceStatuses:
        '*v1.Proxy.gateway-proxy_automtls-istio-edge-api-test':
          reportedBy: gloo
          state: Accepted
```

If the upstream name is the same, the VirtualService could be using an old upstream config and this could be causing the test to pass. 

## API changes
None

## Code changes
None

## CI changes
None

## Docs changes
None

# Context

See slack thread for more context: https://solo-io-corp.slack.com/archives/C03MFATU265/p1716929412244209?thread_ts=1716855588.630339&cid=C03MFATU265 

## Interesting decisions
None

## Testing steps
Ran `TEST_PKG=./test/kubernetes/e2e/tests/  GO_TEST_ARGS="-test.run=TestAutomtlsIstioEdgeApisGateway/IstioIntegration -test.count=5 -test.timeout=25m" make go-test` to check that the test is passing locally consistently. Without this fix, the `TestAutomtlsIstioEdgeApisGateway/IstioIntegration` test would fail pretty consistently one of the five tests when running locally on the TestUpgrade test. 

## Notes for reviewers
None

# Checklist:
